File: drawing-logic/shapes.py
import math
import numpy as np

def rectangle(start_x, start_y, end_x, end_y):
    """
    Generates points for a rectangle and its adjustment points.
    """
    shape_points = [(start_x, start_y), (end_x, start_y), (end_x, end_y), (start_x, end_y), (start_x, start_y)]
    adjust_points = [
        (start_x, start_y),  # Top-left
        (end_x, start_y),    # Top-right
        (end_x, end_y),      # Bottom-right
        (start_x, end_y),    # Bottom-left
        ((start_x + end_x) / 2, start_y), # Top-center
        ((start_x + end_x) / 2, end_y), # Bottom-center
        (start_x, (start_y + end_y) / 2), # Left-center
        (end_x, (start_y + end_y) / 2), # Right-center
        ((start_x + end_x) / 2, (start_y + end_y) / 2) # Center
    ]
    return shape_points, adjust_points

def rotated_rectangle(start_x, start_y, end_x, end_y, angle):
    """
    Generates points for a rotated rectangle and its adjustment points.
    """
    center_x = (start_x + end_x) / 2
    center_y = (start_y + end_y) / 2
    
    points = [
        (start_x, start_y),
        (end_x, start_y),
        (end_x, end_y),
        (start_x, end_y)
    ]
    
    rotated_points = []
    for x, y in points:
        rotated_x = center_x + (x - center_x) * math.cos(angle) - (y - center_y) * math.sin(angle)
        rotated_y = center_y + (x - center_x) * math.sin(angle) + (y - center_y) * math.cos(angle)
        rotated_points.append((rotated_x, rotated_y))
    
    rotated_points.append(rotated_points[0])
    
    adjust_points = []
    for x, y in [
        (start_x, start_y),  # Top-left
        (end_x, start_y),    # Top-right
        (end_x, end_y),      # Bottom-right
        (start_x, end_y),    # Bottom-left
        ((start_x + end_x) / 2, start_y), # Top-center
        ((start_x + end_x) / 2, end_y), # Bottom-center
        (start_x, (start_y + end_y) / 2), # Left-center
        (end_x, (start_y + end_y) / 2), # Right-center
        ((start_x + end_x) / 2, (start_y + end_y) / 2) # Center
    ]:
        rotated_x = center_x + (x - center_x) * math.cos(angle) - (y - center_y) * math.sin(angle)
        rotated_y = center_y + (x - center_x) * math.sin(angle) + (y - center_y) * math.cos(angle)
        adjust_points.append((rotated_x, rotated_y))
    
    return rotated_points, adjust_points

def circle(center_x, center_y, radius, num_points=100):
    """
    Generates points for a circle and its adjustment points.
    """
    shape_points = []
    for i in range(num_points):
        angle = 2 * math.pi * i / num_points
        x = center_x + radius * math.cos(angle)
        y = center_y + radius * math.sin(angle)
        shape_points.append((x, y))
    adjust_points = [
        (center_x + radius, center_y),  # Right
        (center_x - radius, center_y),  # Left
        (center_x, center_y + radius),  # Top
        (center_x, center_y - radius),  # Bottom
        (center_x, center_y) # Center
    ]
    return shape_points, adjust_points

def ellipse(center_x, center_y, radius_x, radius_y, num_points=100):
    """
    Generates points for an ellipse and its adjustment points.
    """
    shape_points = []
    for i in range(num_points):
        angle = 2 * math.pi * i / num_points
        x = center_x + radius_x * math.cos(angle)
        y = center_y + radius_y * math.sin(angle)
        shape_points.append((x, y))
    adjust_points = [
        (center_x + radius_x, center_y),  # Right
        (center_x - radius_x, center_y),  # Left
        (center_x, center_y + radius_y),  # Top
        (center_x, center_y - radius_y),  # Bottom
        (center_x, center_y) # Center
    ]
    return shape_points, adjust_points

def triangle(start_x, start_y, end_x, end_y):
    """
    Generates points for a triangle and its adjustment points.
    """
    mid_x = (start_x + end_x) / 2
    shape_points = [(start_x, end_y), (mid_x, start_y), (end_x, end_y), (start_x, end_y)]
    adjust_points = [
        (start_x, end_y),  # Bottom-left
        (mid_x, start_y),    # Top
        (end_x, end_y),      # Bottom-right
        ((start_x + mid_x) / 2, (start_y + end_y) / 2), # Left-center
        ((mid_x + end_x) / 2, (start_y + end_y) / 2), # Right-center
        ((start_x + end_x) / 2, end_y), # Bottom-center
        (mid_x, (start_y + end_y) / 2) # Center
    ]
    return shape_points, adjust_points

def arc(center_x, center_y, radius, start_angle, end_angle, num_points=100):
    """
    Generates points for an arc and its adjustment points.
    """
    shape_points = []
    angle_range = end_angle - start_angle
    for i in range(num_points):
        angle = start_angle + angle_range * i / num_points
        x = center_x + radius * math.cos(angle)
        y = center_y + radius * math.sin(angle)
        shape_points.append((x, y))
    
    adjust_points = [
        (center_x + radius * math.cos(start_angle), center_y + radius * math.sin(start_angle)), # Start point
        (center_x + radius * math.cos(end_angle), center_y + radius * math.sin(end_angle)), # End point
        (center_x, center_y) # Center
    ]
    return shape_points, adjust_points

def curve(start_x, start_y, control_x, control_y, end_x, end_y, num_points=100):
    """
    Generates points for a quadratic Bezier curve and its adjustment points.
    """
    shape_points = []
    for i in range(num_points + 1):
        t = i / num_points
        x = (1 - t) * (1 - t) * start_x + 2 * (1 - t) * t * control_x + t * t * end_x
        y = (1 - t) * (1 - t) * start_y + 2 * (1 - t) * t * control_y + t * t * end_y
        shape_points.append((x, y))
    adjust_points = [
        (start_x, start_y),
        (control_x, control_y),
        (end_x, end_y)
    ]
    return shape_points, adjust_points

def double_curve(start_x, start_y, control1_x, control1_y, control2_x, control2_y, end_x, end_y, num_points=100):
    """
    Generates points for a cubic Bezier curve and its adjustment points.
    """
    shape_points = []
    for i in range(num_points + 1):
        t = i / num_points
        x = (1 - t)**3 * start_x + 3 * (1 - t)**2 * t * control1_x + 3 * (1 - t) * t**2 * control2_x + t**3 * end_x
        y = (1 - t)**3 * start_y + 3 * (1 - t)**2 * t * control1_y + 3 * (1 - t) * t**2 * control2_y + t**3 * end_y
        shape_points.append((x, y))
    adjust_points = [
        (start_x, start_y),
        (control1_x, control1_y),
        (control2_x, control2_y),
        (end_x, end_y)
    ]
    return shape_points, adjust_points

File: drawing-java/shapes.js
function rectangle(startX, startY, endX, endY) {
    const shapePoints = [
        { x: startX, y: startY },
        { x: endX, y: startY },
        { x: endX, y: endY },
        { x: startX, y: endY },
        { x: startX, y: startY }
    ];

    const adjustPoints = [
        { x: startX, y: startY },
        { x: endX, y: startY },
        { x: endX, y: endY },
        { x: startX, y: endY },
        { x: (startX + endX) / 2, y: startY },
        { x: (startX + endX) / 2, y: endY },
        { x: startX, y: (startY + endY) / 2 },
        { x: endX, y: (startY + endY) / 2 },
        { x: (startX + endX) / 2, y: (startY + endY) / 2 }
    ];

    return { shapePoints, adjustPoints };
}

function rotatedRectangle(startX, startY, endX, endY, angle) {
    const centerX = (startX + endX) / 2;
    const centerY = (startY + endY) / 2;

    const points = [
        { x: startX, y: startY },
        { x: endX, y: startY },
        { x: endX, y: endY },
        { x: startX, y: endY }
    ];

    const rotatedPoints = points.map(point => ({
        x: centerX + (point.x - centerX) * Math.cos(angle) - (point.y - centerY) * Math.sin(angle),
        y: centerY + (point.x - centerX) * Math.sin(angle) + (point.y - centerY) * Math.cos(angle)
    }));

    rotatedPoints.push({ ...rotatedPoints[0] });

    const adjustPoints = rotatedPoints.map(point => ({
        x: point.x,
        y: point.y,
        type: 'adjustment'
    }));

    return { shapePoints: rotatedPoints, adjustPoints };
}

function ellipse(startX, startY, endX, endY, control1X, control1Y, control2X, control2Y) {
    const centerX = (startX + endX) / 2;
    const centerY = (startY + endY) / 2;
    const radiusX = Math.abs(endX - startX) / 2;
    const radiusY = Math.abs(endY - startY) / 2;

    const shapePoints = [];
    const steps = 100;

    for (let i = 0; i <= steps; i++) {
        const angle = (i / steps) * 2 * Math.PI;
        const x = centerX + radiusX * Math.cos(angle);
        const y = centerY + radiusY * Math.sin(angle);
        shapePoints.push({ x, y });
    }

    const adjustPoints = [
        { x: startX, y: startY },
        { x: endX, y: startY },
        { x: endX, y: endY },
        { x: startX, y: endY },
        { x: centerX, y: startY },
        { x: centerX, y: endY },
        { x: startX, y: centerY },
        { x: endX, y: centerY },
        { x: centerX, y: centerY }
    ];

    return { shapePoints, adjustPoints };
}
// ... (keeping existing rectangle, rotatedRectangle, ellipse functions)

function circle(centerX, centerY, radius, numPoints = 100) {
    const shapePoints = [];
    for (let i = 0; i < numPoints; i++) {
        const angle = 2 * Math.PI * i / numPoints;
        const x = centerX + radius * Math.cos(angle);
        const y = centerY + radius * Math.sin(angle);
        shapePoints.push({ x, y });
    }

    const adjustPoints = [
        { x: centerX + radius, y: centerY },
        { x: centerX - radius, y: centerY },
        { x: centerX, y: centerY + radius },
        { x: centerX, y: centerY - radius },
        { x: centerX, y: centerY }
    ];

    return { shapePoints, adjustPoints };
}

function triangle(startX, startY, endX, endY) {
    const midX = (startX + endX) / 2;
    const shapePoints = [
        { x: startX, y: endY },
        { x: midX, y: startY },
        { x: endX, y: endY },
        { x: startX, y: endY }
    ];

    const adjustPoints = [
        { x: startX, y: endY },
        { x: midX, y: startY },
        { x: endX, y: endY },
        { x: (startX + midX) / 2, y: (startY + endY) / 2 },
        { x: (midX + endX) / 2, y: (startY + endY) / 2 },
        { x: (startX + endX) / 2, y: endY },
        { x: midX, y: (startY + endY) / 2 }
    ];

    return { shapePoints, adjustPoints };
}

function arc(centerX, centerY, radius, startAngle, endAngle, numPoints = 100) {
    const shapePoints = [];
    const angleRange = endAngle - startAngle;
    
    for (let i = 0; i < numPoints; i++) {
        const angle = startAngle + angleRange * i / numPoints;
        const x = centerX + radius * Math.cos(angle);
        const y = centerY + radius * Math.sin(angle);
        shapePoints.push({ x, y });
    }

    const adjustPoints = [
        { x: centerX + radius * Math.cos(startAngle), y: centerY + radius * Math.sin(startAngle) },
        { x: centerX + radius * Math.cos(endAngle), y: centerY + radius * Math.sin(endAngle) },
        { x: centerX, y: centerY }
    ];

    return { shapePoints, adjustPoints };
}

function curve(startX, startY, controlX, controlY, endX, endY, numPoints = 100) {
    const shapePoints = [];
    
    for (let i = 0; i <= numPoints; i++) {
        const t = i / numPoints;
        const x = (1 - t) * (1 - t) * startX + 2 * (1 - t) * t * controlX + t * t * endX;
        const y = (1 - t) * (1 - t) * startY + 2 * (1 - t) * t * controlY + t * t * endY;
        shapePoints.push({ x, y });
    }

    const adjustPoints = [
        { x: startX, y: startY },
        { x: controlX, y: controlY },
        { x: endX, y: endY }
    ];

    return { shapePoints, adjustPoints };
}

function doubleCurve(startX, startY, control1X, control1Y, control2X, control2Y, endX, endY, numPoints = 100) {
    const shapePoints = [];
    
    for (let i = 0; i <= numPoints; i++) {
        const t = i / numPoints;
        const x = Math.pow(1 - t, 3) * startX + 
                 3 * Math.pow(1 - t, 2) * t * control1X + 
                 3 * (1 - t) * t * t * control2X + 
                 Math.pow(t, 3) * endX;
        const y = Math.pow(1 - t, 3) * startY + 
                 3 * Math.pow(1 - t, 2) * t * control1Y + 
                 3 * (1 - t) * t * t * control2Y + 
                 Math.pow(t, 3) * endY;
        shapePoints.push({ x, y });
    }

    const adjustPoints = [
        { x: startX, y: startY },
        { x: control1X, y: control1Y },
        { x: control2X, y: control2Y },
        { x: endX, y: endY }
    ];

    return { shapePoints, adjustPoints };
}

export { 
    rectangle, 
    rotatedRectangle, 
    ellipse, 
    circle, 
    triangle, 
    arc, 
    curve, 
    doubleCurve 
};


